### ISSUE TITLE
backend: full support for ESM packages

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#12218]
## Feature Suggestion

Full ESM support, including the package `"exports"` field with the module format specifiers, across all aspects of development.

## Possible Implementation

We currently support ESM in the frontend and tests. Backend support is more tricky as we're executing code in Node.js, which is a lot more picky about module formats. We likely want a combination of build tooling additions and refactors. The optimal solution would ofc be if there's no impact on existing code at all, since in the end our source is mostly ESM.

A couple of dependencies:

- [ ] Support for dynamic imports, #8242
- [x] Support for the exports field, #7926

**Comment by alexandrepesant** [comment]
Hey! Is this something that's being currently looked at? I was building a backend plugin and imported a package that turned out to be built with ESM in mind and couldn't make it work.

**Comment by Rugvip** [comment]
It's not a huge priority at the moment, but we're taking steps towards it. Most recently we added support for the `exports` field in #15816. I think it's best to encourage any upstream package that doesn't build both ESM and CJS versions of their package to do so. There's really no good reason to have node packages that only supports ESM at this point afaik.

**Comment by JounQin** [comment]
@Rugvip 

But some packages have migrated to ESM only, and we can't use them right now.

For example, [node-fetch](https://github.com/node-fetch/node-fetch), [yn](https://github.com/sindresorhus/yn).

**Comment by Rugvip** [comment]
@JounQin Any package can provide both ESM and CJS implementations, if they are only compatible with ESM then that is because they chose to not provide CJS support to their users. Those packages won't be available to use to us and a lot of other users in the Node.js ecosystem. In many cases it's possible to use an older version of the package that still provides CJS support.

**Comment by Rugvip** [comment]
@JounQin yep but ESM doesn't replace CJS, the intention is for both of them to exist as complementary module systems in Node.js. We'll add support for ESM so that we can get the benefits of using that system, when it makes sense to prioritize that work. Until then we'll stick to using packages that provide CJS support.

**Comment by jonkoops** [comment]
Just wanted to chime in here that the reason a lot package maintainers (including myself) are increasingly choosing to adopt 'ESM only' as a strategy, is that we're already spread thin on maintaining a bunch of code, and having both CommonJS and ESM in a package [non trivial](https://blog.isquaredsoftware.com/2023/08/esm-modernization-lessons/).

I really do hope the priorities for this feature are reconsidered as the JavaScript ecosystem as a whole moves on from CommonJS, it would be a shame to get left behind simply because there is no interoperability in Backstage.

**Comment by Rugvip** [comment]
I've dug into this some more again and reached a few conclusions that I'd like some feedback on:

In order to support ESM in backend code we need to add support for it without the need to change **any** existing code. This includes dynamic `import(...)`s and any other types of imports, syntax, or introspection APIs. For this to work we'll need a loader that supports our existing resolution logic during development, as well as a build configuration that is able to transpile our existing code to be valid ESM without a custom loader at runtime. At the TypeScript level I'm hoping we don't need to do anything since nothing has changed, although we'd need to make sure `"typ": "module"` and `.mts` etc. are supported.

The reason for the above approach is that it makes rollout of support for ESM much simpler, but I also think it's a sensible long-term design too. Since Backstage projects often are set up to mix frontend and backend code in the same repository, I think it would be quite confusing and annoying if they used completely different syntax for imports. It could also impact our ability to use dynamic imports in isomorphic `-common` packages.

In the main Backstage projects we still restrict dependencies to only allow usage of packages that are both CJS and ESM compatible. If we don't have this restriction in place, we'd be breaking upstream packages. An exception to this is dynamic imports, which lets us load ESM-only package into CJS. This can be a bit cumbersome to use, but exists as an option.

It is possible that the ecosystem reaches a point where we can move over to ESM-only, but I'm expecting this to be much more long-term. This is reflected in the NPM ecosystem as well, where ESM-only packages still have very low adoption. Take for example `node-fetch`, where the CJS+ESM compatible v2 sits at around 50M weekly downloads, while the new ESM-only v3 is only at around 5M, even though v3 has been out for a long time at this point.

While we shouldn't use ESM-only dependencies in plugin code in the Backstage ecosystem, they can could still be used in end user projects by switching packages or individual modules to ESM. This allows for direct dependencies on ESM-only packages with regular `import ... from ...` syntax.

**Comment by jonkoops** [comment]
> Take for example `node-fetch`, where the CJS+ESM compatible v2 sits at around 50M weekly downloads, while the new ESM-only v3 is only at around 5M, even though v3 has been out for a long time at this point.

To be fair, this is a polyfill. Any version of Node.js in the LTS series has Fetch API support built in these days. Nobody really depends on `node-fetch` directly anymore, it's likely various other package in the ecosystem that pull it in as a transitive dependency.

If you'd like to get an idea of how the adoption of ESM is going then check out [this repository](https://github.com/wooorm/npm-esm-vs-cjs?tab=readme-ov-file#data). It collects the published module variants of the top installed packages on NPM.

**Comment by Rugvip** [comment]
@jonkoops [wooorm/npm-esm-vs-cjs](https://github.com/wooorm/npm-esm-vs-cjs) looks interesting! Only analyzing `latest` seems a bit tricky though, since splitting CJS/ESM support across version seems like a fairly common approach. Another example that might be more representative than `node-fetch` is `p-limit`. Similar situation although for `p-limit` I think it's less of an explicit approach to split support across versions, but instead simply leave the CJS version unmaintained. Still, you end up with a similar split as for `node-fetch` when looking at the usage across versions for the last week:

- CJS: ~107M downloads
- ESM: ~13M downloads

On the topic I'll also highlight that since Backstage packages are more or less TypeScript-only, simultaneous support for btw CJS and ESM.

Still though, if possible I do think it would be much better if we were able to go straight to ESM, I'm just not quite seeing that path 😅

**Comment by jonkoops** [comment]
Indeed, it is often a major version that introduces ESM support, possibly also removing CJS support in the proccess. Libraries that preserve CJS support often have a smoother upgrade path for users, but more maintenance due to the need to introduce a build system, and more code duplication/larger package sizes. It is a tradeoff, and a lot of maintainers whom are overworked open-source hobbiyists are choosing the option that incurs less technical debt on them by going ESM-only, betting on the ecosytem to fix these interop issues.

I helped implement ESM in `jwt-decode` v4, and it preserves CJS support. Still, besides having no real breaking changes besides a slightly different import, [not everyone is moving over](https://www.npmjs.com/package/jwt-decode?activeTab=versions):

- v4  (first published 1 year ago, CJS and ESM): ~2M downloads 
- v3 (first published 4 years ago, CJS only): ~3.6M downloads
- v2 (first published 9 years ago, CJS only): ~1.5M downloads

My takeaway here is that a large part of the JavaScript ecosystem does not upgrade their dependencies, at least not regularly.. ESM is of course a breaking change in that regard that makes adoption more difficult. But even without ESM there is a large tendency not to maintain software.

> Still though, if possible I do think it would be much better if we were able to go straight to ESM, I'm just not quite seeing that path 😅

Would [`require(esm)`](https://joyeecheung.github.io/blog/2024/03/18/require-esm-in-node-js/) help you here? It is an experimental feature, you can read more about it in the [Node.js documentation](https://nodejs.org/api/modules.html#loading-ecmascript-modules-using-require).

**Comment by travi** [comment]
> If we don't have this restriction in place, we'd be breaking upstream packages.

Could you expand on this detail? 

I admit to not knowing the structure of backstage very well since this thread was sent to me by a team that is deeper than I am. The part of the esm topic that is going through my mind here is that a transition to esm is a bit directional. ESM can safely import cjs libraries, but cjs cannot safely import esm libraries without dynamic import. I would expect that the step to make the backstage app esm would be the first step to the journey of making this possible, but I expect that there is more to the structure of the system that I don't have my head around that introduces the risk you are highlighting. 

Could you touch on the imports of the system that present the risk of breaking upstream packages?

**Comment by jonkoops** [comment]
> but cjs cannot safely import esm libraries without dynamic import

That is what `require(esm)` solves. The main problem with ESM is when the module graph is asynchronous, which is incompatible CommonJS on fundamental level. But what was found out that is actually not the case, the ESM graph is only asynchronous if a top-level `await` is used. The blog post by @joyeecheung  that I [linked before](https://joyeecheung.github.io/blog/2024/03/18/require-esm-in-node-js/) is actually a great read to understand the problem.

**Comment by travi** [comment]
> > but cjs cannot safely import esm libraries without dynamic import
> 
> That is what `require(esm)` solves.

great point, and i agree. i debated mentioning that point in addition to dynamic import, but decided against it. the goal of my question was to attempt to understand where esm imports hit that need, whether because the import of an esm module results in a problem, or if it results in a situation where require(esm) could help.

if the base app were esm, it could import esm or cjs modules cleanly. any esm modules imported there could also import either, but cjs imported there requires a change to avoid breakage. what isnt clear to me is which level of the backstage system starts to introduce the situation that would "be breaking upstream packages". 

i fully admit that my understanding lacks clarity around how the backstage structure comes into play in that progression, but I wasnt able to understand that part from this thread. if some clarity around that could be added, i think it could help the conversation be clear about the risks to watch for and either avoid or account for.

**Comment by ddeath** [comment]
I would like to add here some more resources to read on this topic. https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c?permalink_comment_id=3850849#gistcomment-3850849 Looks like guy in the comments is quite knowledgable about the topic. I didnt go through it fully but might be useful for people implementing this support. He has also list of the PRs for other packages he was trying to migrate.

**Comment by jonkoops** [comment]
FYI Node.js version 23 [has been released](https://nodejs.org/en/blog/release/v23.0.0), which includes the ability to import ESM from CommonJS, without any experimental flags. If this continues to work well this feature will be backported to all LTS versions (18, 20, 22).
